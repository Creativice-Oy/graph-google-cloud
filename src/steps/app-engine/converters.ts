import {
  createIntegrationEntity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import { appengine_v1 } from 'googleapis';
import { getGoogleCloudConsoleWebLink } from '../../utils/url';
import {
  ENTITY_CLASS_APP_ENGINE_APPLICATION,
  ENTITY_CLASS_APP_ENGINE_INSTANCE,
  ENTITY_CLASS_APP_ENGINE_SERVICE,
  ENTITY_CLASS_APP_ENGINE_VERSION,
  ENTITY_TYPE_APP_ENGINE_APPLICATION,
  ENTITY_TYPE_APP_ENGINE_INSTANCE,
  ENTITY_TYPE_APP_ENGINE_SERVICE,
  ENTITY_TYPE_APP_ENGINE_VERSION,
} from './constants';

export function createAppEngineApplicationEntity(
  data: appengine_v1.Schema$Application,
  projectId: string,
) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: ENTITY_CLASS_APP_ENGINE_APPLICATION,
        _type: ENTITY_TYPE_APP_ENGINE_APPLICATION,
        _key: data.name as string,
        name: data.name,
        displayName: data.name as string,
        location: data.locationId,
        servingStatus: data.servingStatus,
        defaultHostname: data.defaultHostname,
        defaultBucket: data.defaultBucket,
        splitHealthChecks: data.featureSettings?.splitHealthChecks,
        useContainerOptimizedOs: data.featureSettings?.useContainerOptimizedOs,
        webLink: getGoogleCloudConsoleWebLink(
          `/appengine?project=${projectId}`,
        ),
      },
    },
  });
}

export function createAppEngineServiceEntity(
  data: appengine_v1.Schema$Service,
) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: ENTITY_CLASS_APP_ENGINE_SERVICE,
        _type: ENTITY_TYPE_APP_ENGINE_SERVICE,
        _key: data.name as string,
        name: data.name,
        displayName: data.name as string,
      },
    },
  });
}

export function createAppEngineVersionEntity(
  data: appengine_v1.Schema$Version,
) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: ENTITY_CLASS_APP_ENGINE_VERSION,
        _type: ENTITY_TYPE_APP_ENGINE_VERSION,
        _key: data.name as string,
        name: data.name,
        displayName: data.name as string,
        versionUrl: data.versionUrl,
        category: ['application'],
        createdOn: parseTimePropertyValue(data.createTime),
      },
    },
  });
}

export function createAppEngineInstanceEntity(
  data: appengine_v1.Schema$Instance,
) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: ENTITY_CLASS_APP_ENGINE_INSTANCE,
        _type: ENTITY_TYPE_APP_ENGINE_INSTANCE,
        _key: data.name as string,
        name: data.name,
        displayName: data.name as string,
        availability: data.availability,
        requests: data.requests,
        createdOn: parseTimePropertyValue(data.startTime),
      },
    },
  });
}
