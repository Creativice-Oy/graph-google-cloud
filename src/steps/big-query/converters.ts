import { bigquery_v2 } from 'googleapis';
import { createGoogleCloudIntegrationEntity } from '../../utils/entity';
import {
  BIG_QUERY_DATASET_ENTITY_CLASS,
  BIG_QUERY_DATASET_ENTITY_TYPE,
  BIG_QUERY_TABLE_ENTITY_CLASS,
  BIG_QUERY_TABLE_ENTITY_TYPE,
} from './constants';
import { isMemberPublic } from '../../utils/iam';
import { BigQueryTable } from './client';
import { getGoogleCloudConsoleWebLink } from '../../utils/url';

interface DatasetAccess {
  domain?: string;
  groupByEmail?: string;
  iamMember?: string;
  role?: string;
  specialGroup?: string;
  userByEmail?: string;
}

// 7.1 Ensure that BigQuery datasets are not anonymously or publicly accessible (Scored)
function isBigQueryDatasetPublicAccess(accessList: DatasetAccess[]): boolean {
  for (const access of accessList) {
    const value =
      access.groupByEmail ||
      access.userByEmail ||
      access.iamMember ||
      access.specialGroup ||
      '';
    if (isMemberPublic(value)) {
      return true;
    }
  }

  return false;
}

export function createBigQueryDatasetEntity(data: bigquery_v2.Schema$Dataset) {
  return createGoogleCloudIntegrationEntity(data, {
    entityData: {
      source: data,
      assign: {
        _key: data.id as string,
        _type: BIG_QUERY_DATASET_ENTITY_TYPE,
        _class: BIG_QUERY_DATASET_ENTITY_CLASS,
        name: data.datasetReference?.datasetId,
        description: data.description,
        public: isBigQueryDatasetPublicAccess(data.access || []),
        kmsKeyName: data.defaultEncryptionConfiguration?.kmsKeyName,
        location: data.location,
        encrypted: true,
        classification: null,
        etag: data.etag,
        createdOn: data.creationTime
          ? parseInt(data.creationTime, 10)
          : undefined,
        modifiedOn: data.lastModifiedTime
          ? parseInt(data.lastModifiedTime, 10)
          : undefined,
        webLink: getGoogleCloudConsoleWebLink(
          `/bigquery?project=${data.datasetReference?.projectId}&d=${data.datasetReference?.datasetId}&p=${data.datasetReference?.projectId}&page=dataset`,
        ),
      },
    },
  });
}

export function createBigQueryTableEntity({
  data,
  projectId,
  isPublic,
}: {
  data: BigQueryTable;
  projectId: string;
  isPublic: boolean;
}) {
  return createGoogleCloudIntegrationEntity(data, {
    entityData: {
      source: data,
      assign: {
        _key: data.id as string,
        _type: BIG_QUERY_TABLE_ENTITY_TYPE,
        _class: BIG_QUERY_TABLE_ENTITY_CLASS,
        name: data.tableReference?.tableId,
        public: isPublic,
        type: data.type,
        friendlyName: data.friendlyName,
        classification: null,
        webLink: getGoogleCloudConsoleWebLink(
          `/bigquery?project=${projectId}&d=${data.tableReference?.datasetId}&p=${projectId}&t=${data.tableReference?.tableId}&page=table`,
        ),
        createdOn: data.creationTime
          ? parseInt(data.creationTime, 10)
          : undefined,
        expirationTime: data.expirationTime
          ? parseInt(data.expirationTime, 10)
          : undefined,
      },
    },
  });
}
