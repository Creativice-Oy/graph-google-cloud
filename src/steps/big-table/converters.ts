// import { Entity } from '@jupiterone/integration-sdk-core';
import { bigtableadmin_v2 } from 'googleapis';
import { createGoogleCloudIntegrationEntity } from '../../utils/entity';
import { getGoogleCloudConsoleWebLink } from '../../utils/url';
import {
  ENTITY_CLASS_BIG_TABLE_APP_PROFILE,
  ENTITY_CLASS_BIG_TABLE_BACKUP,
  ENTITY_CLASS_BIG_TABLE_CLUSTER,
  ENTITY_CLASS_BIG_TABLE_INSTANCE,
  ENTITY_CLASS_BIG_TABLE_LOCATION,
  ENTITY_CLASS_BIG_TABLE_OPERATION,
  ENTITY_CLASS_BIG_TABLE_TABLE,
  ENTITY_TYPE_BIG_TABLE_APP_PROFILE,
  ENTITY_TYPE_BIG_TABLE_BACKUP,
  ENTITY_TYPE_BIG_TABLE_CLUSTER,
  ENTITY_TYPE_BIG_TABLE_INSTANCE,
  ENTITY_TYPE_BIG_TABLE_LOCATION,
  ENTITY_TYPE_BIG_TABLE_OPERATION,
  ENTITY_TYPE_BIG_TABLE_TABLE,
} from './constants';

export function getOperationKey(operation: bigtableadmin_v2.Schema$Operation) {
  return `bigtable_operation:${operation.name}`;
}

export function getInstanceKey(instance: bigtableadmin_v2.Schema$Instance) {
  return `bigtable_instance:${instance.name}`;
}

export function getAppProfileKey(
  appProfile: bigtableadmin_v2.Schema$AppProfile,
) {
  return `bigtable_appProfile:${appProfile.name}`;
}

export function getClusterKey(cluster: bigtableadmin_v2.Schema$Cluster) {
  return `bigtable_cluster:${cluster.name}`;
}

export function getBackupKey(backup: bigtableadmin_v2.Schema$Backup) {
  return `bigtable_backup:${backup.name}`;
}

export function getTableKey(table: bigtableadmin_v2.Schema$Table) {
  return `bigtable_table:${table.name}`;
}

export function getLocationKey(location: bigtableadmin_v2.Schema$Location) {
  return `bigtable_location:${location.locationId}`;
}

export function createOperationEntity({
  operation,
}: {
  operation: bigtableadmin_v2.Schema$Operation;
}) {
  return createGoogleCloudIntegrationEntity(operation, {
    entityData: {
      source: operation,
      assign: {
        _class: ENTITY_CLASS_BIG_TABLE_OPERATION,
        _type: ENTITY_TYPE_BIG_TABLE_OPERATION,
        _key: getOperationKey(operation),
        name: operation.name,
        done: operation.done,
      },
    },
  });
}

export function createInstanceEntity({
  instance,
  projectId,
}: {
  instance: bigtableadmin_v2.Schema$Instance;
  projectId: string | undefined | null;
}) {
  return createGoogleCloudIntegrationEntity(instance, {
    entityData: {
      source: instance,
      assign: {
        _class: ENTITY_CLASS_BIG_TABLE_INSTANCE,
        _type: ENTITY_TYPE_BIG_TABLE_INSTANCE,
        _key: getInstanceKey(instance),
        name: instance.name,
        displayName: instance.displayName!,
        state: instance.state,
        type: instance.type,
        webLink: getGoogleCloudConsoleWebLink(
          `/bigtable/instances/${instance.name}/overview?project=${projectId}`,
        ),
      },
    },
  });
}

export function createAppProfileEntity({
  appProfile,
  projectId,
  instanceId,
}: {
  appProfile: bigtableadmin_v2.Schema$AppProfile;
  projectId: string | undefined | null;
  instanceId: string | undefined | null;
}) {
  return createGoogleCloudIntegrationEntity(appProfile, {
    entityData: {
      source: appProfile,
      assign: {
        _class: ENTITY_CLASS_BIG_TABLE_APP_PROFILE,
        _type: ENTITY_TYPE_BIG_TABLE_APP_PROFILE,
        _key: getAppProfileKey(appProfile),
        name: appProfile.name,
        instanceId,
        etag: appProfile.etag,
        description: appProfile.description,
        webLink: getGoogleCloudConsoleWebLink(
          `/bigtable/instances/${instanceId}/app-profiles/${appProfile.name}?project=${projectId}`,
        ),
      },
    },
  });
}

export function createClusterEntity({
  cluster,
  projectId,
  instanceId,
}: {
  cluster: bigtableadmin_v2.Schema$Cluster;
  projectId: string | undefined | null;
  instanceId: string | undefined | null;
}) {
  return createGoogleCloudIntegrationEntity(cluster, {
    entityData: {
      source: cluster,
      assign: {
        _class: ENTITY_CLASS_BIG_TABLE_CLUSTER,
        _type: ENTITY_TYPE_BIG_TABLE_CLUSTER,
        _key: getClusterKey(cluster),
        name: cluster.name,
        instanceId,
        state: cluster.state,
        serveNodes: cluster.serveNodes,
        defaultStorageType: cluster.defaultStorageType,
        location: cluster.location,
        kmsKeyName: cluster.encryptionConfig?.kmsKeyName,
        webLink: getGoogleCloudConsoleWebLink(
          `/bigtable/instances/${instanceId}/overview?project=${projectId}`,
        ),
      },
    },
  });
}

export function createBackupEntity({
  backup,
  projectId,
  instanceId,
  clusterId,
}: {
  backup: bigtableadmin_v2.Schema$Backup;
  projectId: string | undefined | null;
  instanceId: string | undefined | null;
  clusterId: string | undefined | null;
}) {
  return createGoogleCloudIntegrationEntity(backup, {
    entityData: {
      source: backup,
      assign: {
        _class: ENTITY_CLASS_BIG_TABLE_BACKUP,
        _type: ENTITY_TYPE_BIG_TABLE_BACKUP,
        _key: getBackupKey(backup),
        name: backup.name,
        instanceId,
        clusterId,
        sourceTable: backup.sourceTable,
        expireTime: backup.expireTime,
        startTime: backup.startTime,
        endTime: backup.endTime,
        sizeBytes: backup.sizeBytes,
        state: backup.state,
        encryptionType: backup.encryptionInfo?.encryptionType,
        kmsKeyVersion: backup.encryptionInfo?.kmsKeyVersion,
        webLink: getGoogleCloudConsoleWebLink(
          `/bigtable/instances/${instanceId}/backups?project=${projectId}`,
        ),
      },
    },
  });
}

export function createTableEntity({
  table,
  projectId,
  instanceId,
}: {
  table: bigtableadmin_v2.Schema$Table;
  projectId: string | undefined | null;
  instanceId: string | undefined | null;
}) {
  return createGoogleCloudIntegrationEntity(table, {
    entityData: {
      source: table,
      assign: {
        _class: ENTITY_CLASS_BIG_TABLE_TABLE,
        _type: ENTITY_TYPE_BIG_TABLE_TABLE,
        _key: getTableKey(table),
        name: table.name,
        classification: 'unclassified', // not sure what to put here. this is a required field for the class DataCollection
        instanceId,
        granularity: table.granularity || undefined, // these value are there in the type definition in the docs, but are not being sent by the API. we're only getting the name
        sourceType: table.restoreInfo?.sourceType || undefined,
        backup: table.restoreInfo?.backupInfo?.backup || undefined,
        webLink: getGoogleCloudConsoleWebLink(
          `/bigtable/instances/${instanceId}/tables?project=${projectId}`,
        ),
      },
    },
  });
}

export function createLocationEntity({
  location,
}: {
  location: bigtableadmin_v2.Schema$Location;
}) {
  return createGoogleCloudIntegrationEntity(location, {
    entityData: {
      source: location,
      assign: {
        _class: ENTITY_CLASS_BIG_TABLE_LOCATION,
        _type: ENTITY_TYPE_BIG_TABLE_LOCATION,
        _key: getLocationKey(location),
        name: location.name,
        locationId: location.locationId,
        displayName: location.displayName || undefined,
      },
    },
  });
}
