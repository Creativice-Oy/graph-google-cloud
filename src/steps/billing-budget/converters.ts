import { billingbudgets_v1 } from 'googleapis';
import { createGoogleCloudIntegrationEntity } from '../../utils/entity';
import {
  ENTITY_CLASS_BILLING_BUDGET,
  ENTITY_TYPE_BILLING_BUDGET,
} from './constants';

export function createBillingBudgetEntity(
  data: billingbudgets_v1.Schema$GoogleCloudBillingBudgetsV1Budget,
) {
  return createGoogleCloudIntegrationEntity(data, {
    entityData: {
      source: data,
      assign: {
        _class: ENTITY_CLASS_BILLING_BUDGET,
        _type: ENTITY_TYPE_BILLING_BUDGET,
        _key: data.name as string,
        name: data.name,
        displayName: data.displayName as string,
      },
    },
  });
}
