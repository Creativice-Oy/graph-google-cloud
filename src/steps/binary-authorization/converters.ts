import { createIntegrationEntity } from '@jupiterone/integration-sdk-core';
import { binaryauthorization_v1 } from 'googleapis';
import { getGoogleCloudConsoleWebLink } from '../../utils/url';
import {
  BINARY_AUTHORIZATION_POLICY_ENTITY_CLASS,
  BINARY_AUTHORIZATION_POLICY_ENTITY_TYPE,
} from './constants';

export function createBinaryAuthorizationPolicyEntity(
  data: binaryauthorization_v1.Schema$Policy,
  projectId: string,
) {
  const admissionWhitelistPatternsSrc: binaryauthorization_v1.Schema$AdmissionWhitelistPattern[] =
    data.admissionWhitelistPatterns || [];

  const admissionWhitelistPatterns: string[] = admissionWhitelistPatternsSrc
    .map(
      (pattern: binaryauthorization_v1.Schema$AdmissionWhitelistPattern) =>
        pattern.namePattern || '',
    )
    .filter((value) => value);

  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: data.name as string,
        _type: BINARY_AUTHORIZATION_POLICY_ENTITY_TYPE,
        _class: BINARY_AUTHORIZATION_POLICY_ENTITY_CLASS,
        name: data.name,
        // 6.1.4 Minimize Container Registries to only those approved (Not Scored)
        admissionWhitelistPatterns: admissionWhitelistPatterns,
        evaluationMode: data.defaultAdmissionRule?.evaluationMode,
        globalPolicyEvaluationMode: data.globalPolicyEvaluationMode,
        title: 'Binary Authorization Policy',
        summary:
          'Binary Authorization implements a policy model, where a policy is a set of rules that governs the deployment of container images to a GKE cluster. Rules in a policy provide specific criteria that an image must satisfy before it can be deployed.',
        content: '',
        webLink: getGoogleCloudConsoleWebLink(
          `/security/binary-authorization/policy?project=${projectId}&folder=&organizationId=`,
        ),
      },
    },
  });
}
