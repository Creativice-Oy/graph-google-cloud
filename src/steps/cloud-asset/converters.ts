import { cloudasset_v1 } from 'googleapis';
import { snakeCase } from 'lodash';
import { hashArray } from '../../utils/crypto';

import { createGoogleCloudIntegrationEntity } from '../../utils/entity';
import { bindingEntities } from './constants';

export function buildIamBindingEntityKey({
  binding,
  projectName,
  resource,
}: {
  binding: cloudasset_v1.Schema$Binding;
  projectName: string | undefined | null;
  resource: string | undefined | null;
}) {
  // There are no unique values returned from this result so use everything
  // available to make the binding key unique.
  const keyBuilders = ['binding'];

  if (projectName) keyBuilders.push('project:' + projectName);
  if (resource) keyBuilders.push('resource:' + resource);
  if (binding.role) keyBuilders.push('role:' + binding.role);
  if (binding.members)
    keyBuilders.push('members:' + hashArray(binding.members));

  return keyBuilders.join('|');
}

export function createIamBindingEntity({
  _key,
  projectId,
  binding,
  resource,
}: {
  _key: string;
  projectId?: string;
  binding: cloudasset_v1.Schema$Binding;
  resource: string | undefined | null;
}) {
  const namePrefix = 'Role Binding for Resource: ';

  return createGoogleCloudIntegrationEntity(binding, {
    entityData: {
      source: binding,
      assign: {
        _class: bindingEntities.BINDINGS._class,
        _type: bindingEntities.BINDINGS._type,
        _key,
        name: snakeCase(namePrefix) + resource,
        displayName: namePrefix + resource,
        resource,
        role: binding.role,
        members: binding.members,
        projectId,
        'condition.title': binding.condition?.title,
        'condition.description': binding.condition?.description,
        'condition.expression': binding.condition?.expression,
      },
    },
  });
}
