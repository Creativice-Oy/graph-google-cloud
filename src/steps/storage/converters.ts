import { storage_v1 } from 'googleapis';
import { parseTimePropertyValue } from '@jupiterone/integration-sdk-core';
import {
  CLOUD_STORAGE_BUCKET_ENTITY_TYPE,
  CLOUD_STORAGE_BUCKET_ENTITY_CLASS,
} from './constants';
import { createGoogleCloudIntegrationEntity } from '../../utils/entity';
import { NearestOrgPolicyResult, OrgPolicyResult } from '../orgpolicy';

export function getCloudStorageBucketWebLink(
  data: storage_v1.Schema$Bucket,
  projectId: string,
) {
  return `https://console.cloud.google.com/storage/browser/${data.name};tab=objects?forceOnBucketsSortingFiltering=false&project=${projectId}`;
}

export function getCloudStorageBucketKey(id: string) {
  return `bucket:${id}`;
}

function getPublicState(
  accessPreventionPolicy: OrgPolicyResult | undefined,
  isPublic: boolean | undefined,
  isSubjectToObjectAcls: boolean,
) {
  if (accessPreventionPolicy === undefined) {
    // The step responsible for fetching access prevention policies didn't run correctly so we can't know for sure
    return undefined;
  }

  if (
    accessPreventionPolicy.result === NearestOrgPolicyResult.FOUND &&
    accessPreventionPolicy.organizationPolicy?.spec?.rules
  ) {
    if (
      accessPreventionPolicy.organizationPolicy?.spec?.rules[0].enforce === true
    ) {
      // Prevention access policy is found, it's enforced, meaning no bucket is publicly accessible
      return false;
    } else {
      // Prevention access policy is found, but it's not enforced, meaning we fallback to previous code for checking
      return isPublic || isSubjectToObjectAcls;
    }
  }

  if (accessPreventionPolicy.result === NearestOrgPolicyResult.NOT_FOUND) {
    // Prevention access policy wasn't found, meaning we fallback to previous code for checking
    return isPublic || isSubjectToObjectAcls;
  }
}

export function createCloudStorageBucketEntity({
  data,
  projectId,
  isPublic,
  publicAccessPrevention,
}: {
  data: storage_v1.Schema$Bucket;
  projectId: string;
  isPublic?: boolean;
  publicAccessPrevention: OrgPolicyResult | undefined;
}) {
  const isSubjectToObjectAcls =
    data.iamConfiguration?.uniformBucketLevelAccess?.enabled !== true &&
    !isPublic;

  return createGoogleCloudIntegrationEntity(data, {
    entityData: {
      source: data,
      assign: {
        _class: CLOUD_STORAGE_BUCKET_ENTITY_CLASS,
        _type: CLOUD_STORAGE_BUCKET_ENTITY_TYPE,
        _key: getCloudStorageBucketKey(data.id as string),
        id: data.id as string,
        name: data.name,
        displayName: data.name as string,
        storageClass: data.storageClass,
        createdOn: parseTimePropertyValue(data.timeCreated),
        updatedOn: parseTimePropertyValue(data.updated),
        // Storage buckets are encrypted by default
        encrypted: true,
        // If not set, we are using the default Google Encryption key
        encryptionKeyRef: data.encryption?.defaultKmsKeyName,
        kmsKeyName: data.encryption?.defaultKmsKeyName,
        // https://cloud.google.com/storage/docs/uniform-bucket-level-access
        uniformBucketLevelAccess:
          data.iamConfiguration?.uniformBucketLevelAccess?.enabled === true,
        // 2.3 Ensure that retention policies on log buckets are configured using Bucket Lock (Scored)
        retentionPolicyEnabled: data.retentionPolicy?.isLocked,
        retentionPeriod: data.retentionPolicy?.retentionPeriod,
        retentionDate: data.retentionPolicy?.effectiveTime,
        /**
         * It is not possible to know if bucket is public or not when uniformBucketLevelAccess is not enabled
         * as when it is not enabled, each item in the google_storage_bucket is subject to its own Access
         * Control List (ACL) which may or may not be open to the internet. Ingesting the ACLs of every
         * element in a storage bucket is far too large a task for this integration.
         *
         * Ref: https://cloud.google.com/storage/docs/cloud-console?&_ga=2.84754521.-1526178294.1622832983&_gac=1.262728446.1626996208.CjwKCAjwruSHBhAtEiwA_qCppsTtaBT90RDQ-e9xjNnNQM0lwd2aI9wJfUhrVgFjQ0_SDu4kR1yUDhoCeRwQAvD_BwE#_sharingdata
         */
        public: getPublicState(
          publicAccessPrevention,
          isPublic,
          isSubjectToObjectAcls,
        ),
        isSubjectToObjectAcls,
        versioningEnabled: data.versioning?.enabled === true,
        // Rely on the value of the classification tag
        classification: null,
        etag: data.etag,
        webLink: getCloudStorageBucketWebLink(data, projectId),
      },
    },
  });
}
