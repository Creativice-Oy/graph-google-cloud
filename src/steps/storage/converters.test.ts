import { getMockStorageBucket } from '../../../test/mocks';
import { createCloudStorageBucketEntity } from './converters';

const projectId = 'abc123';

describe('#createCloudStorageBucketEntity', () => {
  test('should convert to entity', () => {
    expect(
      createCloudStorageBucketEntity({
        data: getMockStorageBucket(),
        projectId,
        isPublic: false,
      }),
    ).toMatchSnapshot();
  });

  test('should convert to entity with isPublic param set to true', () => {
    expect(
      createCloudStorageBucketEntity({
        data: getMockStorageBucket(),
        projectId,
        isPublic: true,
      }),
    ).toMatchSnapshot();
  });

  test('should convert to entity with uniformBucketAccess enabled', () => {
    expect(
      createCloudStorageBucketEntity({
        data: getMockStorageBucket({
          iamConfiguration: {
            bucketPolicyOnly: {
              enabled: false,
            },
            uniformBucketLevelAccess: {
              enabled: true,
            },
          },
        }),
        projectId,
        isPublic: false,
      }),
    ).toMatchSnapshot();
  });

  test('should handle missing iamConfiguration', () => {
    expect(
      createCloudStorageBucketEntity({
        data: getMockStorageBucket({
          retentionPolicy: {
            retentionPeriod: '600',
            effectiveTime: '2021-01-27T20:10:45.870Z',
            isLocked: true,
          },
        }),
        projectId,
        isPublic: false,
      }),
    ).toMatchSnapshot();
  });

  test('should have retentionPolicyEnabled enabled if bucket is locked', () => {
    expect(
      createCloudStorageBucketEntity({
        data: getMockStorageBucket({
          iamConfiguration: undefined,
        }),
        projectId,
        isPublic: false,
      }),
    ).toMatchSnapshot();
  });
});
