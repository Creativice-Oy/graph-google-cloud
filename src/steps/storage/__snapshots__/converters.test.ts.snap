// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#createCloudStorageBucketEntity should convert to entity 1`] = `
Object {
  "_class": Array [
    "DataStore",
  ],
  "_key": "bucket:customer-managed-encryption-key-bucket-1234",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "encryption": Object {
          "defaultKmsKeyName": "projects/j1-gc-integration-dev/locations/us/keyRings/j1-gc-integration-dev-bucket-ring/cryptoKeys/j1-gc-integration-dev-bucket-key",
        },
        "etag": "CAE=",
        "iamConfiguration": Object {
          "bucketPolicyOnly": Object {
            "enabled": false,
          },
          "uniformBucketLevelAccess": Object {
            "enabled": false,
          },
        },
        "id": "customer-managed-encryption-key-bucket-1234",
        "kind": "storage#bucket",
        "location": "US",
        "locationType": "multi-region",
        "metageneration": "1",
        "name": "customer-managed-encryption-key-bucket-1234",
        "projectNumber": "1234",
        "selfLink": "https://www.googleapis.com/storage/v1/b/customer-managed-encryption-key-bucket-1234",
        "storageClass": "STANDARD",
        "timeCreated": "2020-07-28T19:06:14.033Z",
        "updated": "2020-07-28T19:06:14.033Z",
      },
    },
  ],
  "_type": "google_storage_bucket",
  "classification": null,
  "createdOn": 1595963174033,
  "displayName": "customer-managed-encryption-key-bucket-1234",
  "encrypted": true,
  "encryptionKeyRef": "projects/j1-gc-integration-dev/locations/us/keyRings/j1-gc-integration-dev-bucket-ring/cryptoKeys/j1-gc-integration-dev-bucket-key",
  "etag": "CAE=",
  "id": "customer-managed-encryption-key-bucket-1234",
  "location": "US",
  "name": "customer-managed-encryption-key-bucket-1234",
  "public": false,
  "storageClass": "STANDARD",
  "uniformBucketLevelAccess": false,
  "updatedOn": 1595963174033,
  "webLink": "https://console.cloud.google.com/storage/browser/customer-managed-encryption-key-bucket-1234;tab=objects?forceOnBucketsSortingFiltering=false&project=abc123",
}
`;

exports[`#createCloudStorageBucketEntity should convert to entity with isPublic param set to true 1`] = `
Object {
  "_class": Array [
    "DataStore",
  ],
  "_key": "bucket:customer-managed-encryption-key-bucket-1234",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "encryption": Object {
          "defaultKmsKeyName": "projects/j1-gc-integration-dev/locations/us/keyRings/j1-gc-integration-dev-bucket-ring/cryptoKeys/j1-gc-integration-dev-bucket-key",
        },
        "etag": "CAE=",
        "iamConfiguration": Object {
          "bucketPolicyOnly": Object {
            "enabled": false,
          },
          "uniformBucketLevelAccess": Object {
            "enabled": false,
          },
        },
        "id": "customer-managed-encryption-key-bucket-1234",
        "kind": "storage#bucket",
        "location": "US",
        "locationType": "multi-region",
        "metageneration": "1",
        "name": "customer-managed-encryption-key-bucket-1234",
        "projectNumber": "1234",
        "selfLink": "https://www.googleapis.com/storage/v1/b/customer-managed-encryption-key-bucket-1234",
        "storageClass": "STANDARD",
        "timeCreated": "2020-07-28T19:06:14.033Z",
        "updated": "2020-07-28T19:06:14.033Z",
      },
    },
  ],
  "_type": "google_storage_bucket",
  "classification": null,
  "createdOn": 1595963174033,
  "displayName": "customer-managed-encryption-key-bucket-1234",
  "encrypted": true,
  "encryptionKeyRef": "projects/j1-gc-integration-dev/locations/us/keyRings/j1-gc-integration-dev-bucket-ring/cryptoKeys/j1-gc-integration-dev-bucket-key",
  "etag": "CAE=",
  "id": "customer-managed-encryption-key-bucket-1234",
  "location": "US",
  "name": "customer-managed-encryption-key-bucket-1234",
  "public": true,
  "storageClass": "STANDARD",
  "uniformBucketLevelAccess": false,
  "updatedOn": 1595963174033,
  "webLink": "https://console.cloud.google.com/storage/browser/customer-managed-encryption-key-bucket-1234;tab=objects?forceOnBucketsSortingFiltering=false&project=abc123",
}
`;

exports[`#createCloudStorageBucketEntity should convert to entity with uniformBucketAccess enabled 1`] = `
Object {
  "_class": Array [
    "DataStore",
  ],
  "_key": "bucket:customer-managed-encryption-key-bucket-1234",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "encryption": Object {
          "defaultKmsKeyName": "projects/j1-gc-integration-dev/locations/us/keyRings/j1-gc-integration-dev-bucket-ring/cryptoKeys/j1-gc-integration-dev-bucket-key",
        },
        "etag": "CAE=",
        "iamConfiguration": Object {
          "bucketPolicyOnly": Object {
            "enabled": false,
          },
          "uniformBucketLevelAccess": Object {
            "enabled": true,
          },
        },
        "id": "customer-managed-encryption-key-bucket-1234",
        "kind": "storage#bucket",
        "location": "US",
        "locationType": "multi-region",
        "metageneration": "1",
        "name": "customer-managed-encryption-key-bucket-1234",
        "projectNumber": "1234",
        "selfLink": "https://www.googleapis.com/storage/v1/b/customer-managed-encryption-key-bucket-1234",
        "storageClass": "STANDARD",
        "timeCreated": "2020-07-28T19:06:14.033Z",
        "updated": "2020-07-28T19:06:14.033Z",
      },
    },
  ],
  "_type": "google_storage_bucket",
  "classification": null,
  "createdOn": 1595963174033,
  "displayName": "customer-managed-encryption-key-bucket-1234",
  "encrypted": true,
  "encryptionKeyRef": "projects/j1-gc-integration-dev/locations/us/keyRings/j1-gc-integration-dev-bucket-ring/cryptoKeys/j1-gc-integration-dev-bucket-key",
  "etag": "CAE=",
  "id": "customer-managed-encryption-key-bucket-1234",
  "location": "US",
  "name": "customer-managed-encryption-key-bucket-1234",
  "public": false,
  "storageClass": "STANDARD",
  "uniformBucketLevelAccess": true,
  "updatedOn": 1595963174033,
  "webLink": "https://console.cloud.google.com/storage/browser/customer-managed-encryption-key-bucket-1234;tab=objects?forceOnBucketsSortingFiltering=false&project=abc123",
}
`;

exports[`#createCloudStorageBucketEntity should handle missing iamConfiguration 1`] = `
Object {
  "_class": Array [
    "DataStore",
  ],
  "_key": "bucket:customer-managed-encryption-key-bucket-1234",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "encryption": Object {
          "defaultKmsKeyName": "projects/j1-gc-integration-dev/locations/us/keyRings/j1-gc-integration-dev-bucket-ring/cryptoKeys/j1-gc-integration-dev-bucket-key",
        },
        "etag": "CAE=",
        "iamConfiguration": undefined,
        "id": "customer-managed-encryption-key-bucket-1234",
        "kind": "storage#bucket",
        "location": "US",
        "locationType": "multi-region",
        "metageneration": "1",
        "name": "customer-managed-encryption-key-bucket-1234",
        "projectNumber": "1234",
        "selfLink": "https://www.googleapis.com/storage/v1/b/customer-managed-encryption-key-bucket-1234",
        "storageClass": "STANDARD",
        "timeCreated": "2020-07-28T19:06:14.033Z",
        "updated": "2020-07-28T19:06:14.033Z",
      },
    },
  ],
  "_type": "google_storage_bucket",
  "classification": null,
  "createdOn": 1595963174033,
  "displayName": "customer-managed-encryption-key-bucket-1234",
  "encrypted": true,
  "encryptionKeyRef": "projects/j1-gc-integration-dev/locations/us/keyRings/j1-gc-integration-dev-bucket-ring/cryptoKeys/j1-gc-integration-dev-bucket-key",
  "etag": "CAE=",
  "id": "customer-managed-encryption-key-bucket-1234",
  "location": "US",
  "name": "customer-managed-encryption-key-bucket-1234",
  "public": false,
  "storageClass": "STANDARD",
  "uniformBucketLevelAccess": false,
  "updatedOn": 1595963174033,
  "webLink": "https://console.cloud.google.com/storage/browser/customer-managed-encryption-key-bucket-1234;tab=objects?forceOnBucketsSortingFiltering=false&project=abc123",
}
`;
